Enbedding Model: https://huggingface.co/intfloat/e5-large-v2

This model has 24 layers and the embedding size is 1024.


- Corpus: 13,867 documents (from finder_corpus.jsonl)
- Queries: 218 queries (from finder_queries.jsonl)
- Encoding: The model encodes all documents in 217 batches for the corpus and 4 batches for queries
- Reranking: 675 batches of query-document pairs for cross-encoder reranking

Batching is used for memory efficiency and computational performance:

- Query Encoding (4 batches): Your 128 queries are encoded in batches to avoid GPU/CPU memory overflow
- Corpus Encoding (217 batches): The ~13,900 documents are too large to encode at once, so they're processed in chunks of 64 documents (default batch size)
- Reranking (675 batches): The cross-encoder processes query-document pairs in batches of 32 to compute relevance scores

### Pre-requsite for Milvus DB Impl 

1) Milvus DB running on Docker v2.6.9

![image.png](attachment:image.png)

## Embedding the corpusus and queries using e5-largev2 alrogithm and store them as .pkl file

In [ ]:
"""
Step 1: Encode corpus and queries ONCE using e5-large-v2
Saves embeddings to disk to avoid re-encoding for each index type
"""

import json
import pickle
import numpy as np
from tqdm import tqdm
from sentence_transformers import SentenceTransformer
import os

CORPUS_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/finder_corpus.jsonl/corpus.jsonl"
QUERY_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/finder_queries.jsonl/queries.jsonl"

# Output paths (will be created automatically)
CORPUS_EMBEDDINGS_PATH = "./encoded_data/corpus_embeddings.pkl"
QUERY_EMBEDDINGS_PATH = "./encoded_data/query_embeddings.pkl"
CORPUS_METADATA_PATH = "./encoded_data/corpus_metadata.pkl"
QUERY_METADATA_PATH = "./encoded_data/query_metadata.pkl"

def load_jsonl(path):
    """Load JSONL file"""
    data = []
    with open(path, "r") as f:
        for line in f:
            data.append(json.loads(line))
    return data

def encode_corpus(model, corpus_data, batch_size=32):
    """Encode corpus documents"""
    print(f"\n📚 Encoding {len(corpus_data)} corpus documents...")
    
    # Prepare texts with e5 prompt
    texts = []
    doc_ids = []
    metadata = []
    
    for doc in tqdm(corpus_data, desc="Preparing corpus"):
        doc_id = doc["_id"]
        title = doc.get("title", "")
        text = doc.get("text", "")
        
        # Combine title and text
        full_text = f"{title} {text}".strip()
        
        # Add e5 passage prompt
        texts.append(f"passage: {full_text}")
        doc_ids.append(doc_id)
        metadata.append({
            "doc_id": doc_id,
            "title": title,
            "text": text
        })
    
    # Encode in batches
    print("🔄 Encoding corpus (this will take ~2-3 hours)...")
    embeddings = model.encode(
        texts,
        batch_size=batch_size,
        show_progress_bar=True,
        convert_to_numpy=True,
        normalize_embeddings=True  # Important for cosine similarity
    )
    
    print(f"✅ Corpus encoded: {embeddings.shape}")
    return embeddings, doc_ids, metadata

def encode_queries(model, query_data, batch_size=32):
    """Encode query texts"""
    print(f"\n🔍 Encoding {len(query_data)} queries...")
    
    # Prepare queries with e5 prompt
    texts = []
    query_ids = []
    
    for query in tqdm(query_data, desc="Preparing queries"):
        query_id = query["_id"]
        query_text = query["text"]
        
        # Add e5 query prompt
        texts.append(f"query: {query_text}")
        query_ids.append(query_id)
    
    # Encode
    print("🔄 Encoding queries...")
    embeddings = model.encode(
        texts,
        batch_size=batch_size,
        show_progress_bar=True,
        convert_to_numpy=True,
        normalize_embeddings=True
    )
    
    print(f"✅ Queries encoded: {embeddings.shape}")
    return embeddings, query_ids

def main():
    print("="*60)
    print("Step 1: Encode Data Once (e5-large-v2)")
    print("="*60)
    
    # Create output directory
    os.makedirs("./encoded_data", exist_ok=True)
    
    # Load model
    print("\n📥 Loading e5-large-v2 model...")
    model = SentenceTransformer("intfloat/e5-large-v2")
    print(f"✅ Model loaded. Embedding dimension: {model.get_sentence_embedding_dimension()}")
    
    # Load raw data
    print("\n📂 Loading corpus and queries...")
    corpus_data = load_jsonl(CORPUS_PATH)
    query_data = load_jsonl(QUERY_PATH)
    print(f"✅ Loaded {len(corpus_data)} documents and {len(query_data)} queries")
    
    # Encode corpus
    corpus_embeddings, corpus_ids, corpus_metadata = encode_corpus(model, corpus_data)
    
    # Encode queries
    query_embeddings, query_ids = encode_queries(model, query_data)
    
    # Save to disk
    print("\n💾 Saving embeddings to disk...")
    
    with open(CORPUS_EMBEDDINGS_PATH, 'wb') as f:
        pickle.dump({
            'embeddings': corpus_embeddings,
            'ids': corpus_ids
        }, f)
    
    with open(QUERY_EMBEDDINGS_PATH, 'wb') as f:
        pickle.dump({
            'embeddings': query_embeddings,
            'ids': query_ids
        }, f)
    
    with open(CORPUS_METADATA_PATH, 'wb') as f:
        pickle.dump(corpus_metadata, f)
    
    with open(QUERY_METADATA_PATH, 'wb') as f:
        pickle.dump(query_ids, f)
    
    print("\n✅ All embeddings saved successfully!")
    print(f"\n📊 Summary:")
    print(f"   Corpus embeddings: {corpus_embeddings.shape}")
    print(f"   Query embeddings: {query_embeddings.shape}")
    print(f"   Saved to: ./encoded_data/")

if __name__ == "__main__":
    main()

/opt/homebrew/anaconda3/envs/finrag/lib/python3.11/site-packages/sentence_transformers/cross_encoder/CrossEncoder.py:13: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from tqdm.autonotebook import tqdm, trange


Step 1: Encode Data Once (e5-large-v2)

📥 Loading e5-large-v2 model...
✅ Model loaded. Embedding dimension: 1024

📂 Loading corpus and queries...
✅ Loaded 13867 documents and 216 queries

📚 Encoding 13867 corpus documents...


Preparing corpus: 100%|██████████| 13867/13867 [00:00<00:00, 1742276.42it/s]


🔄 Encoding corpus (this will take ~2-3 hours)...


Batches: 100%|██████████| 434/434 [27:12<00:00,  3.76s/it]


✅ Corpus encoded: (13867, 1024)

🔍 Encoding 216 queries...


Preparing queries: 100%|██████████| 216/216 [00:00<00:00, 329683.28it/s]


🔄 Encoding queries...


Batches: 100%|██████████| 7/7 [00:02<00:00,  3.05it/s]


✅ Queries encoded: (216, 1024)

💾 Saving embeddings to disk...

✅ All embeddings saved successfully!

📊 Summary:
   Corpus embeddings: (13867, 1024)
   Query embeddings: (216, 1024)
   Saved to: ./encoded_data/

✨ You can now run step 2 for each index type WITHOUT re-encoding!


## Using Milvus DB for vector storage with multiple index types. 

### FLAT Algorithm

In [ ]:
"""
Step 2: Benchmark a specific Milvus index type (NOTEBOOK VERSION)
This version can be run directly in Jupyter notebooks

Just change the INDEX_TYPE variable below and run the entire script
"""

import pickle
import time
import json
import pandas as pd
import numpy as np
from pymilvus import (
    connections,
    Collection,
    CollectionSchema,
    FieldSchema,
    DataType,
    utility
)

INDEX_TYPE = "FLAT"  # Options: FLAT, IVF_FLAT, IVF_SQ8, IVF_PQ, HNSW

QRELS_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/FinDER_qrels.tsv"

# Paths (created by step1)
CORPUS_EMBEDDINGS_PATH = "./encoded_data/corpus_embeddings.pkl"
QUERY_EMBEDDINGS_PATH = "./encoded_data/query_embeddings.pkl"

# Milvus connection (Docker standalone)
MILVUS_HOST = "localhost"
MILVUS_PORT = "19530"

# Index configurations
INDEX_CONFIGS = {
    "FLAT": {
        "index_type": "FLAT",
        "metric_type": "IP",
        "params": {}
    },
    "IVF_FLAT": {
        "index_type": "IVF_FLAT",
        "metric_type": "IP",
        "params": {"nlist": 128}
    },
    "IVF_SQ8": {
        "index_type": "IVF_SQ8",
        "metric_type": "IP",
        "params": {"nlist": 128}
    },
    "IVF_PQ": {
        "index_type": "IVF_PQ",
        "metric_type": "IP",
        "params": {"nlist": 128, "m": 16, "nbits": 8}
    },
    "HNSW": {
        "index_type": "HNSW",
        "metric_type": "IP",
        "params": {"M": 16, "efConstruction": 200}
    },
    "DISKANN": {
        "index_type": "DISKANN",
        "metric_type": "IP",
        "params": {}
    }
}

SEARCH_PARAMS = {
    "FLAT": {},
    "IVF_FLAT": {"nprobe": 10},
    "IVF_SQ8": {"nprobe": 10},
    "IVF_PQ": {"nprobe": 10},
    "HNSW": {"ef": 100},
    "DISKANN": {"search_list": 100}
}

def load_embeddings():
    """Load pre-encoded embeddings"""
    print("\n📥 Loading pre-encoded embeddings...")
    
    with open(CORPUS_EMBEDDINGS_PATH, 'rb') as f:
        corpus_data = pickle.load(f)
    
    with open(QUERY_EMBEDDINGS_PATH, 'rb') as f:
        query_data = pickle.load(f)
    
    print(f"✅ Loaded:")
    print(f"   Corpus: {corpus_data['embeddings'].shape}")
    print(f"   Queries: {query_data['embeddings'].shape}")
    
    return corpus_data, query_data

def load_qrels():
    """Load ground truth relevance judgments"""
    df = pd.read_csv(QRELS_PATH, sep="\t")
    qrels = df.groupby("query_id").apply(
        lambda g: dict(zip(g["corpus_id"], g["score"]))
    ).to_dict()
    return qrels

def create_milvus_collection(index_type, embedding_dim=1024):
    """Create Milvus collection with specific index type"""
    
    collection_name = f"finder_{index_type.lower()}"
    
    # Drop if exists
    if utility.has_collection(collection_name):
        print(f"🗑️  Dropping existing collection: {collection_name}")
        utility.drop_collection(collection_name)
    
    # Define schema
    fields = [
        FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=True),
        FieldSchema(name="doc_id", dtype=DataType.VARCHAR, max_length=512),
        FieldSchema(name="embedding", dtype=DataType.FLOAT_VECTOR, dim=embedding_dim)
    ]
    
    schema = CollectionSchema(fields=fields, description=f"FinDER corpus with {index_type} index")
    
    # Create collection
    print(f"📦 Creating collection: {collection_name}")
    collection = Collection(name=collection_name, schema=schema)
    
    return collection

def insert_corpus(collection, corpus_data, batch_size=1000):
    """Insert corpus embeddings into Milvus"""
    
    embeddings = corpus_data['embeddings']
    doc_ids = corpus_data['ids']
    
    print(f"\n📤 Inserting {len(doc_ids)} documents into Milvus...")
    
    # Insert in batches
    for i in range(0, len(doc_ids), batch_size):
        batch_ids = doc_ids[i:i+batch_size]
        batch_embeddings = embeddings[i:i+batch_size].tolist()
        
        entities = [
            batch_ids,
            batch_embeddings
        ]
        
        collection.insert(entities)
        
        if (i + batch_size) % 10000 == 0:
            print(f"   Inserted {min(i+batch_size, len(doc_ids))}/{len(doc_ids)} documents")
    
    print(f"✅ Inserted all {len(doc_ids)} documents")
    
    # Flush to ensure data is persisted
    print("💾 Flushing data to disk...")
    collection.flush()
    print(f"✅ Collection has {collection.num_entities} entities")

def build_index(collection, index_type):
    """Build index on the collection"""
    
    config = INDEX_CONFIGS[index_type]
    
    print(f"\n🔨 Building {index_type} index...")
    print(f"   Config: {config}")
    
    start_time = time.time()
    
    index_params = {
        "index_type": config["index_type"],
        "metric_type": config["metric_type"],
        "params": config["params"]
    }
    
    collection.create_index(
        field_name="embedding",
        index_params=index_params
    )
    
    build_time = time.time() - start_time
    
    print(f"✅ Index built in {build_time:.2f}s")
    
    return build_time

def search_queries(collection, query_data, index_type, top_k=100):
    """Search queries and measure retrieval time"""
    
    embeddings = query_data['embeddings']
    query_ids = query_data['ids']
    
    # Load collection into memory
    print(f"\n🔄 Loading collection into memory...")
    collection.load()
    
    search_params = {
        "metric_type": "IP",
        "params": SEARCH_PARAMS[index_type]
    }
    
    print(f"\n🔍 Searching {len(query_ids)} queries...")
    print(f"   Search params: {search_params}")
    
    results = {}
    query_times = []
    
    # Search each query
    for i, (query_id, query_embedding) in enumerate(zip(query_ids, embeddings)):
        
        start_time = time.time()
        
        search_result = collection.search(
            data=[query_embedding.tolist()],
            anns_field="embedding",
            param=search_params,
            limit=top_k,
            output_fields=["doc_id"]
        )
        
        query_time = time.time() - start_time
        query_times.append(query_time)
        
        # Store results
        results[query_id] = {}
        for j, hit in enumerate(search_result[0]):
            doc_id = hit.entity.get("doc_id")
            score = hit.score
            results[query_id][doc_id] = float(score)
        
        if (i + 1) % 100 == 0:
            avg_time = np.mean(query_times[-100:])
            print(f"   Processed {i+1}/{len(query_ids)} queries (avg: {avg_time*1000:.2f}ms/query)")
    
    total_search_time = sum(query_times)
    avg_query_time = np.mean(query_times)
    
    print(f"\n✅ Search complete:")
    print(f"   Total time: {total_search_time:.2f}s")
    print(f"   Avg per query: {avg_query_time*1000:.2f}ms")
    
    return results, total_search_time, avg_query_time

def calculate_ndcg(qrels, results, k):
    """Calculate NDCG@k"""
    ndcg_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        # Get top-k results
        retrieved = results[query_id]
        sorted_docs = sorted(retrieved.items(), key=lambda x: x[1], reverse=True)[:k]
        
        # Calculate DCG
        dcg = 0.0
        for i, (doc_id, score) in enumerate(sorted_docs):
            rel = relevant_docs.get(doc_id, 0)
            dcg += rel / np.log2(i + 2)
        
        # Calculate IDCG
        ideal_rels = sorted(relevant_docs.values(), reverse=True)[:k]
        idcg = sum(rel / np.log2(i + 2) for i, rel in enumerate(ideal_rels))
        
        # NDCG
        ndcg = dcg / idcg if idcg > 0 else 0
        ndcg_scores.append(ndcg)
    
    return np.mean(ndcg_scores)

def calculate_recall(qrels, results, k):
    """Calculate Recall@k"""
    recall_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = set(list(results[query_id].keys())[:k])
        relevant = set(relevant_docs.keys())
        
        if len(relevant) > 0:
            recall = len(retrieved & relevant) / len(relevant)
            recall_scores.append(recall)
    
    return np.mean(recall_scores)

def evaluate(qrels, results):
    """Evaluate retrieval results"""
    
    print(f"\n📈 Evaluating results...")
    
    metrics = {
        "ndcg@1": calculate_ndcg(qrels, results, 1),
        "ndcg@5": calculate_ndcg(qrels, results, 5),
        "ndcg@10": calculate_ndcg(qrels, results, 10),
        "recall@1": calculate_recall(qrels, results, 1),
        "recall@5": calculate_recall(qrels, results, 5),
        "recall@10": calculate_recall(qrels, results, 10),
    }
    
    print(f"\n📊 Results:")
    for metric, value in metrics.items():
        print(f"   {metric.upper()}: {value:.4f}")
    
    return metrics

def main():
    index_type = INDEX_TYPE
    
    print("="*60)
    print(f"Benchmarking Milvus Index: {index_type}")
    print("="*60)
    
    # Load data
    corpus_data, query_data = load_embeddings()
    qrels = load_qrels()
    
    # Connect to Milvus
    print(f"\n🔌 Connecting to Milvus at {MILVUS_HOST}:{MILVUS_PORT}...")
    connections.connect(host=MILVUS_HOST, port=MILVUS_PORT)
    print("✅ Connected to Milvus")
    
    try:
        # Create collection
        collection = create_milvus_collection(index_type)
        
        # Insert corpus
        insert_corpus(collection, corpus_data)
        
        # Build index
        build_time = build_index(collection, index_type)
        
        # Search queries
        results, search_time, avg_query_time = search_queries(collection, query_data, index_type)
        
        # Evaluate
        metrics = evaluate(qrels, results)
        
        # Save results
        output = {
            "index_type": index_type,
            "index_build_time_sec": round(build_time, 2),
            "total_search_time_sec": round(search_time, 2),
            "avg_query_time_ms": round(avg_query_time * 1000, 2),
            **{k: round(v, 4) for k, v in metrics.items()}
        }
        
        output_file = f"./results_{index_type.lower()}.json"
        with open(output_file, 'w') as f:
            json.dump(output, f, indent=2)
        
        print(f"\n💾 Results saved to: {output_file}")
        
        # Return results for notebook display
        return output
        
    finally:
        # Cleanup
        connections.disconnect("default")
        print("\n🔌 Disconnected from Milvus")

if __name__ == "__main__":
    results = main()

Benchmarking Milvus Index: FLAT

📥 Loading pre-encoded embeddings...
✅ Loaded:
   Corpus: (13867, 1024)
   Queries: (216, 1024)

🔌 Connecting to Milvus at localhost:19530...
✅ Connected to Milvus
📦 Creating collection: finder_flat

📤 Inserting 13867 documents into Milvus...


/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_17364/2031576184.py:98: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  qrels = df.groupby("query_id").apply(


   Inserted 10000/13867 documents
✅ Inserted all 13867 documents
💾 Flushing data to disk...
✅ Collection has 13867 entities

🔨 Building FLAT index...
   Config: {'index_type': 'FLAT', 'metric_type': 'IP', 'params': {}}
✅ Index built in 0.53s

🔄 Loading collection into memory...

🔍 Searching 216 queries...
   Search params: {'metric_type': 'IP', 'params': {}}
   Processed 100/216 queries (avg: 6.84ms/query)
   Processed 200/216 queries (avg: 4.83ms/query)

✅ Search complete:
   Total time: 1.25s
   Avg per query: 5.80ms

📈 Evaluating results...

📊 Results:
   NDCG@1: 0.3125
   NDCG@5: 0.3860
   NDCG@10: 0.4281
   RECALL@1: 0.2531
   RECALL@5: 0.4594
   RECALL@10: 0.5742

💾 Results saved to: ./results_flat.json

🔌 Disconnected from Milvus


### Sample Milvus DB Collections Fields/Schema

![image.png](attachment:image.png)

# Run All Benchmarks

In [ ]:
"""
Run All Benchmarks - Jupyter Notebook Version
Tests all 6 Milvus index types and displays comprehensive comparison

Just run this entire script in your Jupyter notebook!
"""

import pickle
import time
import json
import pandas as pd
import numpy as np
import os
from pymilvus import (
    connections,
    Collection,
    CollectionSchema,
    FieldSchema,
    DataType,
    utility
)

QRELS_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/FinDER_qrels.tsv"

# Paths (created by step1)
CORPUS_EMBEDDINGS_PATH = "./encoded_data/corpus_embeddings.pkl"
QUERY_EMBEDDINGS_PATH = "./encoded_data/query_embeddings.pkl"

# Milvus connection
MILVUS_HOST = "localhost"
MILVUS_PORT = "19530"

# Index types to test
INDEX_TYPES = ["FLAT", "IVF_FLAT", "IVF_SQ8", "IVF_PQ", "HNSW", "DISKANN"]

# Index configurations
INDEX_CONFIGS = {
    "FLAT": {
        "index_type": "FLAT",
        "metric_type": "IP",
        "params": {}
    },
    "IVF_FLAT": {
        "index_type": "IVF_FLAT",
        "metric_type": "IP",
        "params": {"nlist": 128}
    },
    "IVF_SQ8": {
        "index_type": "IVF_SQ8",
        "metric_type": "IP",
        "params": {"nlist": 128}
    },
    "IVF_PQ": {
        "index_type": "IVF_PQ",
        "metric_type": "IP",
        "params": {"nlist": 128, "m": 16, "nbits": 8}
    },
    "HNSW": {
        "index_type": "HNSW",
        "metric_type": "IP",
        "params": {"M": 16, "efConstruction": 200}
    },
    "DISKANN": {
        "index_type": "DISKANN",
        "metric_type": "IP",
        "params": {"M": 16, "efConstruction": 200}
    }
}

SEARCH_PARAMS = {
    "FLAT": {},
    "IVF_FLAT": {"nprobe": 10},
    "IVF_SQ8": {"nprobe": 10},
    "IVF_PQ": {"nprobe": 10},
    "HNSW": {"ef": 100},
    "DISKANN": {"search_list": 100}
}

def load_embeddings():
    """Load pre-encoded embeddings"""
    with open(CORPUS_EMBEDDINGS_PATH, 'rb') as f:
        corpus_data = pickle.load(f)
    
    with open(QUERY_EMBEDDINGS_PATH, 'rb') as f:
        query_data = pickle.load(f)
    
    return corpus_data, query_data

def load_qrels():
    """Load ground truth relevance judgments"""
    df = pd.read_csv(QRELS_PATH, sep="\t")
    qrels = df.groupby("query_id").apply(
        lambda g: dict(zip(g["corpus_id"], g["score"]))
    ).to_dict()
    return qrels

def create_milvus_collection(index_type, embedding_dim=1024):
    """Create Milvus collection"""
    collection_name = f"finder_{index_type.lower()}"
    
    if utility.has_collection(collection_name):
        utility.drop_collection(collection_name)
    
    fields = [
        FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=True),
        FieldSchema(name="doc_id", dtype=DataType.VARCHAR, max_length=512),
        FieldSchema(name="embedding", dtype=DataType.FLOAT_VECTOR, dim=embedding_dim)
    ]
    
    schema = CollectionSchema(fields=fields, description=f"FinDER corpus with {index_type} index")
    collection = Collection(name=collection_name, schema=schema)
    
    return collection

def insert_corpus(collection, corpus_data, batch_size=1000):
    """Insert corpus embeddings into Milvus"""
    embeddings = corpus_data['embeddings']
    doc_ids = corpus_data['ids']
    
    for i in range(0, len(doc_ids), batch_size):
        batch_ids = doc_ids[i:i+batch_size]
        batch_embeddings = embeddings[i:i+batch_size].tolist()
        
        entities = [batch_ids, batch_embeddings]
        collection.insert(entities)
    
    collection.flush()

def build_index(collection, index_type):
    """Build index on the collection"""
    config = INDEX_CONFIGS[index_type]
    
    start_time = time.time()
    
    index_params = {
        "index_type": config["index_type"],
        "metric_type": config["metric_type"],
        "params": config["params"]
    }
    
    collection.create_index(field_name="embedding", index_params=index_params)
    
    build_time = time.time() - start_time
    return build_time

def search_queries(collection, query_data, index_type, top_k=100):
    """Search queries and measure retrieval time"""
    embeddings = query_data['embeddings']
    query_ids = query_data['ids']
    
    collection.load()
    
    search_params = {
        "metric_type": "IP",
        "params": SEARCH_PARAMS[index_type]
    }
    
    results = {}
    query_times = []
    
    for i, (query_id, query_embedding) in enumerate(zip(query_ids, embeddings)):
        start_time = time.time()
        
        search_result = collection.search(
            data=[query_embedding.tolist()],
            anns_field="embedding",
            param=search_params,
            limit=top_k,
            output_fields=["doc_id"]
        )
        
        query_time = time.time() - start_time
        query_times.append(query_time)
        
        results[query_id] = {}
        for j, hit in enumerate(search_result[0]):
            doc_id = hit.entity.get("doc_id")
            score = hit.score
            results[query_id][doc_id] = float(score)
    
    total_search_time = sum(query_times)
    avg_query_time = np.mean(query_times)
    
    return results, total_search_time, avg_query_time

def calculate_ndcg(qrels, results, k):
    """Calculate NDCG@k"""
    ndcg_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = results[query_id]
        sorted_docs = sorted(retrieved.items(), key=lambda x: x[1], reverse=True)[:k]
        
        dcg = 0.0
        for i, (doc_id, score) in enumerate(sorted_docs):
            rel = relevant_docs.get(doc_id, 0)
            dcg += rel / np.log2(i + 2)
        
        ideal_rels = sorted(relevant_docs.values(), reverse=True)[:k]
        idcg = sum(rel / np.log2(i + 2) for i, rel in enumerate(ideal_rels))
        
        ndcg = dcg / idcg if idcg > 0 else 0
        ndcg_scores.append(ndcg)
    
    return np.mean(ndcg_scores)

def calculate_recall(qrels, results, k):
    """Calculate Recall@k"""
    recall_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = set(list(results[query_id].keys())[:k])
        relevant = set(relevant_docs.keys())
        
        if len(relevant) > 0:
            recall = len(retrieved & relevant) / len(relevant)
            recall_scores.append(recall)
    
    return np.mean(recall_scores)

def calculate_precision(qrels, results, k):
    """Calculate Precision@k"""
    precision_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = list(results[query_id].keys())[:k]
        relevant = set(relevant_docs.keys())
        
        if len(retrieved) > 0:
            precision = len([doc for doc in retrieved if doc in relevant]) / len(retrieved)
            precision_scores.append(precision)
    
    return np.mean(precision_scores)

def calculate_map(qrels, results, k):
    """Calculate MAP@k (Mean Average Precision)"""
    ap_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = list(results[query_id].keys())[:k]
        relevant = set(relevant_docs.keys())
        
        if len(relevant) == 0:
            continue
        
        precision_sum = 0.0
        num_relevant = 0
        
        for i, doc_id in enumerate(retrieved):
            if doc_id in relevant:
                num_relevant += 1
                precision_at_i = num_relevant / (i + 1)
                precision_sum += precision_at_i
        
        ap = precision_sum / len(relevant) if len(relevant) > 0 else 0
        ap_scores.append(ap)
    
    return np.mean(ap_scores)

def evaluate(qrels, results):
    """Evaluate retrieval results with comprehensive metrics"""
    metrics = {
        "ndcg@1": calculate_ndcg(qrels, results, 1),
        "ndcg@5": calculate_ndcg(qrels, results, 5),
        "ndcg@10": calculate_ndcg(qrels, results, 10),
        "recall@1": calculate_recall(qrels, results, 1),
        "recall@5": calculate_recall(qrels, results, 5),
        "recall@10": calculate_recall(qrels, results, 10),
        "precision@1": calculate_precision(qrels, results, 1),
        "precision@5": calculate_precision(qrels, results, 5),
        "precision@10": calculate_precision(qrels, results, 10),
        "map@10": calculate_map(qrels, results, 10),
    }
    
    return metrics

def benchmark_index(index_type, corpus_data, query_data, qrels):
    """Benchmark a single index type"""
    print(f"\n{'='*60}")
    print(f"Testing: {index_type}")
    print(f"{'='*60}")
    
    try:
        # Connect to Milvus
        connections.connect(host=MILVUS_HOST, port=MILVUS_PORT)
        
        # Create collection
        print(f"📦 Creating collection...")
        collection = create_milvus_collection(index_type)
        
        # Insert corpus
        print(f"📤 Inserting {len(corpus_data['ids'])} documents...")
        insert_corpus(collection, corpus_data)
        
        # Build index
        print(f"🔨 Building {index_type} index...")
        build_time = build_index(collection, index_type)
        print(f"✅ Index built in {build_time:.2f}s")
        
        # Search queries
        print(f"🔍 Searching {len(query_data['ids'])} queries...")
        results, search_time, avg_query_time = search_queries(collection, query_data, index_type)
        print(f"✅ Search complete: {search_time:.2f}s total, {avg_query_time*1000:.2f}ms avg")
        
        # Evaluate
        print(f"📈 Evaluating...")
        metrics = evaluate(qrels, results)
        
        # Prepare output
        output = {
            "index_type": index_type,
            "index_build_time_sec": round(build_time, 2),
            "total_search_time_sec": round(search_time, 2),
            "avg_query_time_ms": round(avg_query_time * 1000, 2),
            **{k: round(v, 4) for k, v in metrics.items()}
        }
        
        # Save individual result
        os.makedirs("./results", exist_ok=True)
        output_file = f"./results/results_{index_type.lower()}.json"
        with open(output_file, 'w') as f:
            json.dump(output, f, indent=2)
        
        print(f"✅ {index_type} Complete!")
        print(f"   NDCG@10: {metrics['ndcg@10']:.4f}")
        print(f"   Recall@10: {metrics['recall@10']:.4f}")
        print(f"   Precision@10: {metrics['precision@10']:.4f}")
        print(f"   MAP@10: {metrics['map@10']:.4f}")
        
        connections.disconnect("default")
        
        return output
        
    except Exception as e:
        print(f"❌ ERROR with {index_type}: {e}")
        connections.disconnect("default")
        return None

def display_results(all_results):
    """Display comprehensive comparison table"""
    print("\n" + "="*120)
    print("MILVUS INDEX BENCHMARK SUMMARY - COMPREHENSIVE METRICS")
    print("="*120)
    
    # Create DataFrame for better display
    df = pd.DataFrame(all_results)
    
    # Reorder columns for better readability
    column_order = [
        'index_type',
        'index_build_time_sec',
        'total_search_time_sec',
        'avg_query_time_ms',
        'ndcg@10',
        'recall@10',
        'precision@10',
        'map@10',
        'ndcg@1',
        'ndcg@5',
        'recall@1',
        'recall@5',
        'precision@1',
        'precision@5'
    ]
    
    df = df[column_order]
    
    # Display summary table (key metrics)
    print("\n📊 KEY PERFORMANCE METRICS:")
    print("-"*120)
    summary_df = df[['index_type', 'index_build_time_sec', 'total_search_time_sec', 
                      'avg_query_time_ms', 'ndcg@10', 'recall@10', 'precision@10', 'map@10']]
    print(summary_df.to_string(index=False))
    
    # Display detailed accuracy metrics
    print("\n\n📈 DETAILED ACCURACY METRICS:")
    print("-"*120)
    accuracy_df = df[['index_type', 'ndcg@1', 'ndcg@5', 'ndcg@10', 
                       'recall@1', 'recall@5', 'recall@10',
                       'precision@1', 'precision@5', 'precision@10', 'map@10']]
    print(accuracy_df.to_string(index=False))
    
    # Find best performers
    print("\n\n🏆 BEST PERFORMERS:")
    print("-"*120)
    print(f"Fastest Search:     {df.loc[df['avg_query_time_ms'].idxmin(), 'index_type']} "
          f"({df['avg_query_time_ms'].min():.2f}ms per query)")
    print(f"Highest NDCG@10:    {df.loc[df['ndcg@10'].idxmax(), 'index_type']} "
          f"({df['ndcg@10'].max():.4f})")
    print(f"Highest Recall@10:  {df.loc[df['recall@10'].idxmax(), 'index_type']} "
          f"({df['recall@10'].max():.4f})")
    print(f"Highest Precision@10: {df.loc[df['precision@10'].idxmax(), 'index_type']} "
          f"({df['precision@10'].max():.4f})")
    print(f"Highest MAP@10:     {df.loc[df['map@10'].idxmax(), 'index_type']} "
          f"({df['map@10'].max():.4f})")
    print(f"Fastest Index Build: {df.loc[df['index_build_time_sec'].idxmin(), 'index_type']} "
          f"({df['index_build_time_sec'].min():.2f}s)")
    
    # Recommendations
    print("\n\n💡 RECOMMENDATIONS:")
    print("-"*120)
    
    # Find balanced performer (high accuracy + fast search)
    df['score'] = (df['ndcg@10'] * 0.4 + df['recall@10'] * 0.3 + 
                   (1 - df['avg_query_time_ms'] / df['avg_query_time_ms'].max()) * 0.3)
    best_balanced = df.loc[df['score'].idxmax(), 'index_type']
    
    print(f"Best Overall (Speed + Accuracy): {best_balanced}")
    print(f"For Production Use: HNSW (usually best balance)")
    print(f"For Perfect Accuracy: FLAT (slowest but 100% accurate)")
    print(f"For Memory Constrained: IVF_SQ8 (compressed)")
    print(f"For Massive Scale: DISKANN (disk-based)")
    
    return df

def main():
    """Run all benchmarks"""
    print("="*60)
    print("MILVUS INDEX BENCHMARKING SUITE")
    print("="*60)
    print(f"\n📋 Will test {len(INDEX_TYPES)} index types:")
    for idx in INDEX_TYPES:
        print(f"   - {idx}")
    
    # Load data once
    print("\n📥 Loading pre-encoded embeddings...")
    corpus_data, query_data = load_embeddings()
    qrels = load_qrels()
    print(f"✅ Loaded {corpus_data['embeddings'].shape[0]} documents and {query_data['embeddings'].shape[0]} queries")
    
    # Run benchmarks
    all_results = []
    
    for index_type in INDEX_TYPES:
        result = benchmark_index(index_type, corpus_data, query_data, qrels)
        if result:
            all_results.append(result)
        
        # Wait between tests
        if index_type != INDEX_TYPES[-1]:
            print("\n⏳ Waiting 10 seconds before next test...")
            time.sleep(10)
    
    # Save combined results
    with open("./results/combined_results.json", 'w') as f:
        json.dump(all_results, f, indent=2)
    
    print("\n💾 All results saved to ./results/")
    
    # Display comprehensive comparison
    df = display_results(all_results)
    
    print("\n\n🎉 All benchmarks complete!")
    
    return df

# Run the benchmarks
if __name__ == "__main__":
    results_df = main()

MILVUS INDEX BENCHMARKING SUITE

📋 Will test 6 index types:
   - FLAT
   - IVF_FLAT
   - IVF_SQ8
   - IVF_PQ
   - HNSW
   - DISKANN

📥 Loading pre-encoded embeddings...
✅ Loaded 13867 documents and 216 queries

Testing: FLAT
📦 Creating collection...


/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_17364/709128494.py:93: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  qrels = df.groupby("query_id").apply(


📤 Inserting 13867 documents...
🔨 Building FLAT index...
✅ Index built in 0.52s
🔍 Searching 216 queries...
✅ Search complete: 1.18s total, 5.47ms avg
📈 Evaluating...
✅ FLAT Complete!
   NDCG@10: 0.4281
   Recall@10: 0.5742
   Precision@10: 0.0797
   MAP@10: 0.3685

⏳ Waiting 10 seconds before next test...

Testing: IVF_FLAT
📦 Creating collection...
📤 Inserting 13867 documents...
🔨 Building IVF_FLAT index...
✅ Index built in 1.02s
🔍 Searching 216 queries...
✅ Search complete: 0.62s total, 2.89ms avg
📈 Evaluating...
✅ IVF_FLAT Complete!
   NDCG@10: 0.4260
   Recall@10: 0.5664
   Precision@10: 0.0797
   MAP@10: 0.3672

⏳ Waiting 10 seconds before next test...

Testing: IVF_SQ8
📦 Creating collection...
📤 Inserting 13867 documents...
🔨 Building IVF_SQ8 index...
✅ Index built in 1.03s
🔍 Searching 216 queries...
✅ Search complete: 0.55s total, 2.53ms avg
📈 Evaluating...
✅ IVF_SQ8 Complete!
   NDCG@10: 0.4260
   Recall@10: 0.5664
   Precision@10: 0.0797
   MAP@10: 0.3672

⏳ Waiting 10 seconds b

| Index Type | Build Time (s) | Total Search (s) | Avg Query (ms) | NDCG@1 | NDCG@5 | NDCG@10 | Recall@1 | Recall@5 | Recall@10 | Prec@1 | Prec@5 | Prec@10 | MAP@10 |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| **FLAT** | 0.52 | 1.18 | 5.47 | 0.3125 | 0.3860 | 0.4281 | 0.2531 | 0.4594 | 0.5742 | 0.3125 | 0.1219 | 0.0797 | 0.3685 |
| **IVF_FLAT** | 1.02 | 0.62 | 2.89 | 0.3125 | 0.3890 | 0.4260 | 0.2531 | 0.4672 | 0.5664 | 0.3125 | 0.1250 | 0.0797 | 0.3672 |
| **IVF_SQ8** | 1.03 | 0.55 | 2.53 | 0.3125 | 0.3890 | 0.4260 | 0.2531 | 0.4672 | 0.5664 | 0.3125 | 0.1250 | 0.0797 | 0.3672 |
| **IVF_PQ** | 3.56 | 0.54 | 2.50 | 0.1562 | 0.2206 | 0.2514 | 0.1289 | 0.2740 | 0.3604 | 0.1562 | 0.0781 | 0.0547 | 0.2015 |
| **HNSW** | 2.04 | 0.59 | 2.75 | 0.3125 | 0.3761 | 0.4183 | 0.2531 | 0.4438 | 0.5586 | 0.3125 | 0.1188 | 0.0781 | 0.3607 |
| **DISKANN** | 11.64 | 1.07 | 4.96 | 0.3125 | 0.3860 | 0.4281 | 0.2531 | 0.4594 | 0.5742 | 0.3125 | 0.1219 | 0.0797 | 0.3685 |

1) **Best Accuracy:** FLAT and DISKANN share the highest accuracy metrics (NDCG, Recall, and MAP), though DISKANN takes significantly longer to build.

2) **Fastest Query Speed:** IVF_PQ and IVF_SQ8 offer the lowest average query times (~2.5ms), though IVF_PQ shows a significant drop in search quality.

3) **Balanced Performer:** IVF_SQ8 provides a great middle ground, offering nearly half the latency of FLAT with almost identical accuracy.

## Postgress Vector DB

In [1]:
"""
PostgreSQL pgvector Benchmark - Complete Implementation
Reuses the same .pkl files created by Milvus encoding step

Tests 3 PostgreSQL index types:
1. No Index (Sequential Scan) - Like Milvus FLAT
2. IVF (Inverted File) - Like Milvus IVF_FLAT  
3. HNSW (Hierarchical Navigable Small World) - Like Milvus HNSW

Run this in Jupyter notebook or as a Python script
"""

import pickle
import time
import json
import pandas as pd
import numpy as np
import psycopg2
from psycopg2.extras import execute_values
import os

# ⚠️ UPDATE THESE PATHS ⚠️
QRELS_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/FinDER_qrels.tsv"
POSTGRES_CONNECTION = "postgresql://vikashpr@localhost:5432/financerag"

# Reuse Milvus encoded data (same .pkl files!)
CORPUS_EMBEDDINGS_PATH = "./encoded_data/corpus_embeddings.pkl"
QUERY_EMBEDDINGS_PATH = "./encoded_data/query_embeddings.pkl"

# PostgreSQL index configurations
INDEX_CONFIGS = {
    "NO_INDEX": {
        "name": "No Index (Sequential Scan)",
        "description": "Exact search, scans all vectors (like Milvus FLAT)",
        "index_sql": None,  # No index created
        "search_type": "Sequential"
    },
    "IVF": {
        "name": "IVF Index",
        "description": "Inverted File index (like Milvus IVF_FLAT)",
        "index_sql": "CREATE INDEX {index_name} ON {table_name} USING ivfflat (embedding vector_cosine_ops) WITH (lists = 100);",
        "search_type": "IVF",
        "params": {"lists": 100, "probes": 10}
    },
    "HNSW": {
        "name": "HNSW Index",
        "description": "Graph-based fast search (like Milvus HNSW)",
        "index_sql": "CREATE INDEX {index_name} ON {table_name} USING hnsw (embedding vector_cosine_ops) WITH (m = 16, ef_construction = 64);",
        "search_type": "HNSW",
        "params": {"m": 16, "ef_construction": 64, "ef_search": 40}
    }
}

def load_embeddings():
    """Load pre-encoded embeddings (same files as Milvus!)"""
    print("\n📥 Loading pre-encoded embeddings...")
    
    with open(CORPUS_EMBEDDINGS_PATH, 'rb') as f:
        corpus_data = pickle.load(f)
    
    with open(QUERY_EMBEDDINGS_PATH, 'rb') as f:
        query_data = pickle.load(f)
    
    print(f"✅ Loaded:")
    print(f"   Corpus: {corpus_data['embeddings'].shape}")
    print(f"   Queries: {query_data['embeddings'].shape}")
    
    return corpus_data, query_data

def load_qrels():
    """Load ground truth relevance judgments"""
    df = pd.read_csv(QRELS_PATH, sep="\t")
    qrels = df.groupby("query_id").apply(
        lambda g: dict(zip(g["corpus_id"], g["score"]))
    ).to_dict()
    return qrels

def create_table(conn, table_name, embedding_dim=1024):
    """Create PostgreSQL table for embeddings"""
    cur = conn.cursor()
    
    # Drop if exists
    cur.execute(f"DROP TABLE IF EXISTS {table_name} CASCADE;")
    
    # Create table
    cur.execute(f"""
        CREATE TABLE {table_name} (
            id SERIAL PRIMARY KEY,
            doc_id TEXT NOT NULL,
            embedding vector({embedding_dim}) NOT NULL
        );
    """)
    
    conn.commit()
    cur.close()
    print(f"✅ Created table: {table_name}")

def insert_corpus(conn, table_name, corpus_data, batch_size=1000):
    """Insert corpus embeddings into PostgreSQL"""
    embeddings = corpus_data['embeddings']
    doc_ids = corpus_data['ids']
    
    print(f"\n📤 Inserting {len(doc_ids)} documents into PostgreSQL...")
    
    cur = conn.cursor()
    
    # Prepare data for batch insert
    data = []
    for doc_id, embedding in zip(doc_ids, embeddings):
        # Convert numpy array to list for PostgreSQL
        data.append((doc_id, embedding.tolist()))
    
    # Insert in batches
    for i in range(0, len(data), batch_size):
        batch = data[i:i+batch_size]
        execute_values(
            cur,
            f"INSERT INTO {table_name} (doc_id, embedding) VALUES %s",
            batch,
            template="(%s, %s)"
        )
        
        if (i + batch_size) % 5000 == 0:
            print(f"   Inserted {min(i+batch_size, len(data))}/{len(data)} documents")
    
    conn.commit()
    cur.close()
    
    print(f"✅ Inserted all {len(doc_ids)} documents")

def build_index(conn, table_name, index_type):
    """Build index on the table"""
    config = INDEX_CONFIGS[index_type]
    
    if config["index_sql"] is None:
        print(f"\n⚪ No index (Sequential Scan)")
        return 0.0
    
    print(f"\n🔨 Building {index_type} index...")
    print(f"   Description: {config['description']}")
    
    cur = conn.cursor()
    
    # Set work_mem for index building (helps with performance)
    cur.execute("SET maintenance_work_mem = '2GB';")
    
    start_time = time.time()
    
    index_name = f"{table_name}_{index_type.lower()}_idx"
    index_sql = config["index_sql"].format(
        index_name=index_name,
        table_name=table_name
    )
    
    cur.execute(index_sql)
    conn.commit()
    
    build_time = time.time() - start_time
    
    cur.close()
    
    print(f"✅ Index built in {build_time:.2f}s")
    
    return build_time

def search_queries(conn, table_name, query_data, index_type, top_k=100):
    """Search queries and measure retrieval time"""
    embeddings = query_data['embeddings']
    query_ids = query_data['ids']
    
    config = INDEX_CONFIGS[index_type]
    
    # Set search parameters for IVF
    cur = conn.cursor()
    if index_type == "IVF":
        probes = config["params"]["probes"]
        cur.execute(f"SET ivfflat.probes = {probes};")
    elif index_type == "HNSW":
        ef_search = config["params"]["ef_search"]
        cur.execute(f"SET hnsw.ef_search = {ef_search};")
    
    print(f"\n🔍 Searching {len(query_ids)} queries...")
    if index_type != "NO_INDEX":
        print(f"   Search params: {config.get('params', {})}")
    
    results = {}
    query_times = []
    
    # Search each query
    for i, (query_id, query_embedding) in enumerate(zip(query_ids, embeddings)):
        
        start_time = time.time()
        
        # PostgreSQL cosine distance query
        # Note: pgvector uses <=> for cosine distance (lower is better)
        # We convert to similarity (1 - distance) to match Milvus
        cur.execute(f"""
            SELECT doc_id, 1 - (embedding <=> %s::vector) as score
            FROM {table_name}
            ORDER BY embedding <=> %s::vector
            LIMIT %s
        """, (query_embedding.tolist(), query_embedding.tolist(), top_k))
        
        query_time = time.time() - start_time
        query_times.append(query_time)
        
        # Store results
        results[query_id] = {}
        for doc_id, score in cur.fetchall():
            results[query_id][doc_id] = float(score)
        
        if (i + 1) % 50 == 0:
            avg_time = np.mean(query_times[-50:])
            print(f"   Processed {i+1}/{len(query_ids)} queries (avg: {avg_time*1000:.2f}ms/query)")
    
    cur.close()
    
    total_search_time = sum(query_times)
    avg_query_time = np.mean(query_times)
    
    print(f"\n✅ Search complete:")
    print(f"   Total time: {total_search_time:.2f}s")
    print(f"   Avg per query: {avg_query_time*1000:.2f}ms")
    
    return results, total_search_time, avg_query_time

def calculate_ndcg(qrels, results, k):
    """Calculate NDCG@k"""
    ndcg_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = results[query_id]
        sorted_docs = sorted(retrieved.items(), key=lambda x: x[1], reverse=True)[:k]
        
        dcg = 0.0
        for i, (doc_id, score) in enumerate(sorted_docs):
            rel = relevant_docs.get(doc_id, 0)
            dcg += rel / np.log2(i + 2)
        
        ideal_rels = sorted(relevant_docs.values(), reverse=True)[:k]
        idcg = sum(rel / np.log2(i + 2) for i, rel in enumerate(ideal_rels))
        
        ndcg = dcg / idcg if idcg > 0 else 0
        ndcg_scores.append(ndcg)
    
    return np.mean(ndcg_scores)

def calculate_recall(qrels, results, k):
    """Calculate Recall@k"""
    recall_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = set(list(results[query_id].keys())[:k])
        relevant = set(relevant_docs.keys())
        
        if len(relevant) > 0:
            recall = len(retrieved & relevant) / len(relevant)
            recall_scores.append(recall)
    
    return np.mean(recall_scores)

def calculate_precision(qrels, results, k):
    """Calculate Precision@k"""
    precision_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = list(results[query_id].keys())[:k]
        relevant = set(relevant_docs.keys())
        
        if len(retrieved) > 0:
            precision = len([doc for doc in retrieved if doc in relevant]) / len(retrieved)
            precision_scores.append(precision)
    
    return np.mean(precision_scores)

def calculate_map(qrels, results, k):
    """Calculate MAP@k"""
    ap_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = list(results[query_id].keys())[:k]
        relevant = set(relevant_docs.keys())
        
        if len(relevant) == 0:
            continue
        
        precision_sum = 0.0
        num_relevant = 0
        
        for i, doc_id in enumerate(retrieved):
            if doc_id in relevant:
                num_relevant += 1
                precision_at_i = num_relevant / (i + 1)
                precision_sum += precision_at_i
        
        ap = precision_sum / len(relevant) if len(relevant) > 0 else 0
        ap_scores.append(ap)
    
    return np.mean(ap_scores)

def evaluate(qrels, results):
    """Evaluate retrieval results"""
    print(f"\n📈 Evaluating...")
    
    metrics = {
        "ndcg@1": calculate_ndcg(qrels, results, 1),
        "ndcg@5": calculate_ndcg(qrels, results, 5),
        "ndcg@10": calculate_ndcg(qrels, results, 10),
        "recall@1": calculate_recall(qrels, results, 1),
        "recall@5": calculate_recall(qrels, results, 5),
        "recall@10": calculate_recall(qrels, results, 10),
        "precision@1": calculate_precision(qrels, results, 1),
        "precision@5": calculate_precision(qrels, results, 5),
        "precision@10": calculate_precision(qrels, results, 10),
        "map@10": calculate_map(qrels, results, 10),
    }
    
    return metrics

def benchmark_index(index_type, corpus_data, query_data, qrels):
    """Benchmark a single PostgreSQL index type"""
    print(f"\n{'='*60}")
    print(f"Testing: {index_type}")
    print(f"{'='*60}")
    
    table_name = f"finder_{index_type.lower()}"
    
    try:
        # Connect to PostgreSQL
        conn = psycopg2.connect(POSTGRES_CONNECTION)
        conn.autocommit = False
        
        # Create table
        create_table(conn, table_name)
        
        # Insert corpus
        insert_corpus(conn, table_name, corpus_data)
        
        # Build index
        build_time = build_index(conn, table_name, index_type)
        
        # Search queries
        results, search_time, avg_query_time = search_queries(
            conn, table_name, query_data, index_type
        )
        
        # Evaluate
        metrics = evaluate(qrels, results)
        
        # Prepare output
        output = {
            "index_type": index_type,
            "database": "PostgreSQL",
            "index_build_time_sec": round(build_time, 2),
            "total_search_time_sec": round(search_time, 2),
            "avg_query_time_ms": round(avg_query_time * 1000, 2),
            **{k: round(v, 4) for k, v in metrics.items()}
        }
        
        # Save individual result
        os.makedirs("./results", exist_ok=True)
        output_file = f"./results/postgres_results_{index_type.lower()}.json"
        with open(output_file, 'w') as f:
            json.dump(output, f, indent=2)
        
        print(f"\n✅ {index_type} Complete!")
        print(f"   NDCG@10: {metrics['ndcg@10']:.4f}")
        print(f"   Recall@10: {metrics['recall@10']:.4f}")
        print(f"   Precision@10: {metrics['precision@10']:.4f}")
        print(f"   MAP@10: {metrics['map@10']:.4f}")
        
        conn.close()
        
        return output
        
    except Exception as e:
        print(f"❌ ERROR with {index_type}: {e}")
        import traceback
        traceback.print_exc()
        try:
            conn.close()
        except:
            pass
        return None

def display_results(postgres_results):
    """Display PostgreSQL results"""
    print("\n" + "="*100)
    print("POSTGRESQL PGVECTOR BENCHMARK SUMMARY")
    print("="*100)
    
    df = pd.DataFrame(postgres_results)
    
    # Key metrics
    print("\n📊 KEY PERFORMANCE METRICS:")
    print("-"*100)
    summary_df = df[['index_type', 'index_build_time_sec', 'total_search_time_sec',
                      'avg_query_time_ms', 'ndcg@10', 'recall@10', 'precision@10', 'map@10']]
    print(summary_df.to_string(index=False))
    
    # Best performers
    print("\n\n🏆 BEST PERFORMERS:")
    print("-"*100)
    print(f"Fastest Search:      {df.loc[df['avg_query_time_ms'].idxmin(), 'index_type']} "
          f"({df['avg_query_time_ms'].min():.2f}ms per query)")
    print(f"Highest NDCG@10:     {df.loc[df['ndcg@10'].idxmax(), 'index_type']} "
          f"({df['ndcg@10'].max():.4f})")
    print(f"Highest Recall@10:   {df.loc[df['recall@10'].idxmax(), 'index_type']} "
          f"({df['recall@10'].max():.4f})")
    
    return df

def main():
    """Run all PostgreSQL benchmarks"""
    print("="*60)
    print("POSTGRESQL PGVECTOR BENCHMARKING SUITE")
    print("="*60)
    print("\n📌 Reusing embeddings from Milvus encoding step!")
    print(f"📋 Will test {len(INDEX_CONFIGS)} index types:")
    for idx, config in INDEX_CONFIGS.items():
        print(f"   - {idx}: {config['description']}")
    
    # Check PostgreSQL connection
    try:
        conn = psycopg2.connect(POSTGRES_CONNECTION)
        cur = conn.cursor()
        cur.execute("SELECT extname FROM pg_extension WHERE extname = 'vector';")
        if not cur.fetchone():
            print("\n❌ ERROR: pgvector extension not installed!")
            print("   Run in psql: CREATE EXTENSION vector;")
            return
        cur.close()
        conn.close()
        print("✅ PostgreSQL connection OK, pgvector installed")
    except Exception as e:
        print(f"\n❌ ERROR: Cannot connect to PostgreSQL: {e}")
        return
    
    # Load data
    corpus_data, query_data = load_embeddings()
    qrels = load_qrels()
    print(f"✅ Loaded {len(qrels)} query judgments")
    
    # Run benchmarks
    all_results = []
    
    for index_type in INDEX_CONFIGS.keys():
        result = benchmark_index(index_type, corpus_data, query_data, qrels)
        if result:
            all_results.append(result)
        
        # Wait between tests
        if index_type != list(INDEX_CONFIGS.keys())[-1]:
            print("\n⏳ Waiting 5 seconds before next test...")
            time.sleep(5)
    
    # Save combined results
    with open("./results/postgres_combined_results.json", 'w') as f:
        json.dump(all_results, f, indent=2)
    
    print("\n💾 All PostgreSQL results saved to ./results/")
    
    # Display comparison
    df = display_results(all_results)
    
    print("\n\n🎉 All PostgreSQL benchmarks complete!")
    
    return df

if __name__ == "__main__":
    results_df = main()

POSTGRESQL PGVECTOR BENCHMARKING SUITE

📌 Reusing embeddings from Milvus encoding step!
📋 Will test 3 index types:
   - NO_INDEX: Exact search, scans all vectors (like Milvus FLAT)
   - IVF: Inverted File index (like Milvus IVF_FLAT)
   - HNSW: Graph-based fast search (like Milvus HNSW)
✅ PostgreSQL connection OK, pgvector installed

📥 Loading pre-encoded embeddings...
✅ Loaded:
   Corpus: (13867, 1024)
   Queries: (216, 1024)
✅ Loaded 64 query judgments

Testing: NO_INDEX
✅ Created table: finder_no_index

📤 Inserting 13867 documents into PostgreSQL...


/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_68419/2599032946.py:73: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  qrels = df.groupby("query_id").apply(


   Inserted 5000/13867 documents
   Inserted 10000/13867 documents
✅ Inserted all 13867 documents

⚪ No index (Sequential Scan)

🔍 Searching 216 queries...
   Processed 50/216 queries (avg: 55.75ms/query)
   Processed 100/216 queries (avg: 55.31ms/query)
   Processed 150/216 queries (avg: 55.52ms/query)
   Processed 200/216 queries (avg: 56.04ms/query)

✅ Search complete:
   Total time: 12.02s
   Avg per query: 55.66ms

📈 Evaluating...

✅ NO_INDEX Complete!
   NDCG@10: 0.4281
   Recall@10: 0.5742
   Precision@10: 0.0797
   MAP@10: 0.3685

⏳ Waiting 5 seconds before next test...

Testing: IVF
✅ Created table: finder_ivf

📤 Inserting 13867 documents into PostgreSQL...
   Inserted 5000/13867 documents
   Inserted 10000/13867 documents
✅ Inserted all 13867 documents

🔨 Building IVF index...
   Description: Inverted File index (like Milvus IVF_FLAT)
✅ Index built in 4.98s

🔍 Searching 216 queries...
   Search params: {'lists': 100, 'probes': 10}
   Processed 50/216 queries (avg: 8.55ms/quer

| Index Type | Database | Build Time (s) | Total Search (s) | Avg Query (ms) | NDCG@1 | NDCG@10 | Recall@1 | Recall@10 | Prec@1 | Prec@10 | MAP@10 |
| :--- | :--- | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| **FLAT** | Native | 0.52 | 1.18 | 5.47 | 0.3125 | 0.4281 | 0.2531 | 0.5742 | 0.3125 | 0.0797 | 0.3685 |
| **IVF_FLAT** | Native | 1.02 | 0.62 | 2.89 | 0.3125 | 0.4260 | 0.2531 | 0.5664 | 0.3125 | 0.0797 | 0.3672 |
| **IVF_SQ8** | Native | 1.03 | 0.55 | 2.53 | 0.3125 | 0.4260 | 0.2531 | 0.5664 | 0.3125 | 0.0797 | 0.3672 |
| **IVF_PQ** | Native | 3.56 | 0.54 | 2.50 | 0.1562 | 0.2514 | 0.1289 | 0.3604 | 0.1562 | 0.0547 | 0.2015 |
| **HNSW** | Native | 2.04 | 0.59 | 2.75 | 0.3125 | 0.4183 | 0.2531 | 0.5586 | 0.3125 | 0.0781 | 0.3607 |
| **DISKANN** | Native | 11.64 | 1.07 | 4.96 | 0.3125 | 0.4281 | 0.2531 | 0.5742 | 0.3125 | 0.0797 | 0.3685 |
| **NO_INDEX** | PostgreSQL | 0.00 | 12.02 | 55.66 | 0.3125 | 0.4281 | 0.2531 | 0.5742 | 0.3125 | 0.0797 | 0.3685 |
| **IVF** | PostgreSQL | 4.98 | 1.76 | 8.17 | 0.3125 | 0.4226 | 0.2531 | 0.5664 | 0.3125 | 0.0797 | 0.3630 |
| **HNSW** | PostgreSQL | 15.01 | 1.45 | 6.70 | 0.2969 | 0.4027 | 0.2375 | 0.5430 | 0.2969 | 0.0766 | 0.3451 |

## OpenSearch Vector DB

### Pre-requisites for OpenSearch Implementation

1) OpenSearch running on Docker with k-NN plugin enabled
2) Connection URL: `http://localhost:9200` (default)

**OpenSearch k-NN Index Types to Test:**
- **NO_INDEX (Script Scoring)** - Exact search, like Milvus FLAT
- **FAISS_HNSW** - FAISS engine with HNSW algorithm
- **FAISS_IVF** - FAISS engine with IVF algorithm  
- **NMSLIB_HNSW** - nmslib engine with HNSW algorithm
- **LUCENE_HNSW** - Lucene engine with HNSW algorithm

**Docker command to start OpenSearch:**
```bash
docker run -d --name opensearch \
  -p 9200:9200 -p 9600:9600 \
  -e "discovery.type=single-node" \
  -e "OPENSEARCH_INITIAL_ADMIN_PASSWORD=Admin@123" \
  opensearchproject/opensearch:2.11.0
```

In [7]:
"""
OpenSearch k-NN Benchmark - Complete Implementation
Reuses the same .pkl files created by Milvus encoding step

Tests 5 OpenSearch index types:
1. No Index (Script Scoring) - Exact search like Milvus FLAT
2. FAISS HNSW - FAISS engine with HNSW algorithm
3. FAISS IVF - FAISS engine with IVF algorithm
4. NMSLIB HNSW - nmslib engine with HNSW algorithm
5. LUCENE HNSW - Lucene engine with HNSW algorithm

Run this in Jupyter notebook or as a Python script
"""

import pickle
import time
import json
import pandas as pd
import numpy as np
from opensearchpy import OpenSearch, helpers
import os

# ⚠️ UPDATE THESE PATHS ⚠️
QRELS_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/FinDER_qrels.tsv"

# OpenSearch connection settings
OPENSEARCH_HOST = "localhost"
OPENSEARCH_PORT = 9200
OPENSEARCH_USER = "admin"
OPENSEARCH_PASSWORD = "Admin@123"  # Update with your password

# Reuse Milvus encoded data (same .pkl files!)
CORPUS_EMBEDDINGS_PATH = "./encoded_data/corpus_embeddings.pkl"
QUERY_EMBEDDINGS_PATH = "./encoded_data/query_embeddings.pkl"

# OpenSearch k-NN index configurations
INDEX_CONFIGS = {
    "NO_INDEX": {
        "name": "No Index (Script Scoring)",
        "description": "Exact search using script scoring (like Milvus FLAT)",
        "use_script_scoring": True,
        "knn_settings": None
    },
    "FAISS_HNSW": {
        "name": "FAISS HNSW Index",
        "description": "FAISS engine with HNSW algorithm",
        "use_script_scoring": False,
        "knn_settings": {
            "type": "knn_vector",
            "dimension": 1024,
            "method": {
                "name": "hnsw",
                "space_type": "cosinesimil",
                "engine": "faiss",
                "parameters": {
                    "ef_construction": 128,
                    "m": 16
                }
            }
        },
        "search_params": {"ef_search": 100}
    },
    "FAISS_IVF": {
        "name": "FAISS IVF Index",
        "description": "FAISS engine with IVF algorithm",
        "use_script_scoring": False,
        "knn_settings": {
            "type": "knn_vector",
            "dimension": 1024,
            "method": {
                "name": "ivf",
                "space_type": "cosinesimil",
                "engine": "faiss",
                "parameters": {
                    "nlist": 100,
                    "nprobes": 10
                }
            }
        },
        "search_params": {"nprobes": 10}
    },
    "NMSLIB_HNSW": {
        "name": "NMSLIB HNSW Index",
        "description": "nmslib engine with HNSW algorithm",
        "use_script_scoring": False,
        "knn_settings": {
            "type": "knn_vector",
            "dimension": 1024,
            "method": {
                "name": "hnsw",
                "space_type": "cosinesimil",
                "engine": "nmslib",
                "parameters": {
                    "ef_construction": 128,
                    "m": 16
                }
            }
        },
        "search_params": {"ef_search": 100}
    },
    "LUCENE_HNSW": {
        "name": "Lucene HNSW Index",
        "description": "Lucene engine with HNSW algorithm",
        "use_script_scoring": False,
        "knn_settings": {
            "type": "knn_vector",
            "dimension": 1024,
            "method": {
                "name": "hnsw",
                "space_type": "cosinesimil",
                "engine": "lucene",
                "parameters": {
                    "ef_construction": 128,
                    "m": 16
                }
            }
        },
        "search_params": {"ef_search": 100}
    }
}


def get_opensearch_client():
    """Create OpenSearch client connection"""
    # For security disabled mode (DISABLE_SECURITY_PLUGIN=true), use_ssl=False
    # For security enabled mode, use_ssl=True with verify_certs=False
    client = OpenSearch(
        hosts=[{"host": OPENSEARCH_HOST, "port": OPENSEARCH_PORT}],
        http_auth=(OPENSEARCH_USER, OPENSEARCH_PASSWORD),
        use_ssl=False,  # Set to False when DISABLE_SECURITY_PLUGIN=true
        verify_certs=False,
        ssl_show_warn=False,
        timeout=300
    )
    return client


def load_embeddings():
    """Load pre-encoded embeddings (same files as Milvus!)"""
    print("\n📥 Loading pre-encoded embeddings...")
    
    with open(CORPUS_EMBEDDINGS_PATH, 'rb') as f:
        corpus_data = pickle.load(f)
    
    with open(QUERY_EMBEDDINGS_PATH, 'rb') as f:
        query_data = pickle.load(f)
    
    print(f"✅ Loaded:")
    print(f"   Corpus: {corpus_data['embeddings'].shape}")
    print(f"   Queries: {query_data['embeddings'].shape}")
    
    return corpus_data, query_data


def load_qrels():
    """Load ground truth relevance judgments"""
    df = pd.read_csv(QRELS_PATH, sep="\t")
    qrels = df.groupby("query_id").apply(
        lambda g: dict(zip(g["corpus_id"], g["score"]))
    ).to_dict()
    return qrels


def create_index(client, index_name, index_type):
    """Create OpenSearch index with k-NN settings"""
    config = INDEX_CONFIGS[index_type]
    
    # Delete existing index
    if client.indices.exists(index=index_name):
        client.indices.delete(index=index_name)
        print(f"   Deleted existing index: {index_name}")
    
    if config["use_script_scoring"]:
        # For script scoring, create a simple index without k-NN
        index_body = {
            "settings": {
                "number_of_shards": 1,
                "number_of_replicas": 0
            },
            "mappings": {
                "properties": {
                    "doc_id": {"type": "keyword"},
                    "embedding": {
                        "type": "dense_vector",
                        "dims": 1024,
                        "index": False  # No indexing for exact search
                    }
                }
            }
        }
    else:
        # For k-NN indexing
        index_body = {
            "settings": {
                "index": {
                    "knn": True,
                    "knn.algo_param.ef_search": config.get("search_params", {}).get("ef_search", 100)
                },
                "number_of_shards": 1,
                "number_of_replicas": 0
            },
            "mappings": {
                "properties": {
                    "doc_id": {"type": "keyword"},
                    "embedding": config["knn_settings"]
                }
            }
        }
    
    client.indices.create(index=index_name, body=index_body)
    print(f"✅ Created index: {index_name}")


def insert_corpus(client, index_name, corpus_data, batch_size=500):
    """Insert corpus embeddings into OpenSearch"""
    embeddings = corpus_data['embeddings']
    doc_ids = corpus_data['ids']
    
    print(f"\n📤 Inserting {len(doc_ids)} documents into OpenSearch...")
    
    # Prepare bulk actions
    def generate_actions():
        for i, (doc_id, embedding) in enumerate(zip(doc_ids, embeddings)):
            yield {
                "_index": index_name,
                "_id": str(i),
                "_source": {
                    "doc_id": doc_id,
                    "embedding": embedding.tolist()
                }
            }
            if (i + 1) % 5000 == 0:
                print(f"   Prepared {i+1}/{len(doc_ids)} documents")
    
    # Bulk insert
    success, failed = helpers.bulk(
        client,
        generate_actions(),
        chunk_size=batch_size,
        request_timeout=300,
        raise_on_error=False
    )
    
    print(f"✅ Inserted {success} documents (failed: {failed if isinstance(failed, int) else len(failed)})")
    
    # Refresh index to make documents searchable
    client.indices.refresh(index=index_name)
    print("✅ Index refreshed")


def build_index_and_wait(client, index_name, index_type):
    """Wait for index to be fully built"""
    config = INDEX_CONFIGS[index_type]
    
    if config["use_script_scoring"]:
        print(f"\n⚪ No k-NN index (Script Scoring mode)")
        return 0.0
    
    print(f"\n🔨 Building {index_type} index...")
    print(f"   Description: {config['description']}")
    
    start_time = time.time()
    
    # Force merge to build the k-NN graph
    client.indices.forcemerge(index=index_name, max_num_segments=1)
    
    # Wait for index to be ready
    while True:
        stats = client.indices.stats(index=index_name)
        doc_count = stats['indices'][index_name]['primaries']['docs']['count']
        
        health = client.cluster.health(index=index_name)
        if health['status'] in ['yellow', 'green']:
            break
        
        time.sleep(1)
    
    build_time = time.time() - start_time
    
    print(f"✅ Index built in {build_time:.2f}s")
    
    return build_time


def search_with_knn(client, index_name, query_embedding, top_k=100):
    """Search using k-NN query"""
    query = {
        "size": top_k,
        "query": {
            "knn": {
                "embedding": {
                    "vector": query_embedding.tolist(),
                    "k": top_k
                }
            }
        }
    }
    
    response = client.search(index=index_name, body=query)
    return response


def search_with_script_scoring(client, index_name, query_embedding, top_k=100):
    """Search using script scoring (exact search)"""
    query = {
        "size": top_k,
        "query": {
            "script_score": {
                "query": {"match_all": {}},
                "script": {
                    "source": "cosineSimilarity(params.query_vector, 'embedding') + 1.0",
                    "params": {
                        "query_vector": query_embedding.tolist()
                    }
                }
            }
        }
    }
    
    response = client.search(index=index_name, body=query)
    return response


def search_queries(client, index_name, query_data, index_type, top_k=100):
    """Search queries and measure retrieval time"""
    embeddings = query_data['embeddings']
    query_ids = query_data['ids']
    
    config = INDEX_CONFIGS[index_type]
    use_script = config["use_script_scoring"]
    
    print(f"\n🔍 Searching {len(query_ids)} queries...")
    print(f"   Method: {'Script Scoring (exact)' if use_script else 'k-NN Index'}")
    
    results = {}
    query_times = []
    
    # Search each query
    for i, (query_id, query_embedding) in enumerate(zip(query_ids, embeddings)):
        
        start_time = time.time()
        
        try:
            if use_script:
                response = search_with_script_scoring(client, index_name, query_embedding, top_k)
            else:
                response = search_with_knn(client, index_name, query_embedding, top_k)
            
            query_time = time.time() - start_time
            query_times.append(query_time)
            
            # Store results
            results[query_id] = {}
            for hit in response['hits']['hits']:
                doc_id = hit['_source']['doc_id']
                score = hit['_score']
                results[query_id][doc_id] = float(score)
        
        except Exception as e:
            print(f"   ⚠️ Error on query {i}: {e}")
            query_times.append(0)
            results[query_id] = {}
        
        if (i + 1) % 50 == 0:
            avg_time = np.mean(query_times[-50:])
            print(f"   Processed {i+1}/{len(query_ids)} queries (avg: {avg_time*1000:.2f}ms/query)")
    
    total_search_time = sum(query_times)
    avg_query_time = np.mean(query_times)
    
    print(f"\n✅ Search complete:")
    print(f"   Total time: {total_search_time:.2f}s")
    print(f"   Avg per query: {avg_query_time*1000:.2f}ms")
    
    return results, total_search_time, avg_query_time


def calculate_ndcg(qrels, results, k):
    """Calculate NDCG@k"""
    ndcg_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = results[query_id]
        sorted_docs = sorted(retrieved.items(), key=lambda x: x[1], reverse=True)[:k]
        
        dcg = 0.0
        for i, (doc_id, score) in enumerate(sorted_docs):
            rel = relevant_docs.get(doc_id, 0)
            dcg += rel / np.log2(i + 2)
        
        ideal_rels = sorted(relevant_docs.values(), reverse=True)[:k]
        idcg = sum(rel / np.log2(i + 2) for i, rel in enumerate(ideal_rels))
        
        ndcg = dcg / idcg if idcg > 0 else 0
        ndcg_scores.append(ndcg)
    
    return np.mean(ndcg_scores)


def calculate_recall(qrels, results, k):
    """Calculate Recall@k"""
    recall_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = set(list(results[query_id].keys())[:k])
        relevant = set(relevant_docs.keys())
        
        if len(relevant) > 0:
            recall = len(retrieved & relevant) / len(relevant)
            recall_scores.append(recall)
    
    return np.mean(recall_scores)


def calculate_precision(qrels, results, k):
    """Calculate Precision@k"""
    precision_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = list(results[query_id].keys())[:k]
        relevant = set(relevant_docs.keys())
        
        if len(retrieved) > 0:
            precision = len([doc for doc in retrieved if doc in relevant]) / len(retrieved)
            precision_scores.append(precision)
    
    return np.mean(precision_scores)


def calculate_map(qrels, results, k):
    """Calculate MAP@k"""
    ap_scores = []
    
    for query_id, relevant_docs in qrels.items():
        if query_id not in results:
            continue
        
        retrieved = list(results[query_id].keys())[:k]
        relevant = set(relevant_docs.keys())
        
        if len(relevant) == 0:
            continue
        
        precision_sum = 0.0
        num_relevant = 0
        
        for i, doc_id in enumerate(retrieved):
            if doc_id in relevant:
                num_relevant += 1
                precision_at_i = num_relevant / (i + 1)
                precision_sum += precision_at_i
        
        ap = precision_sum / len(relevant) if len(relevant) > 0 else 0
        ap_scores.append(ap)
    
    return np.mean(ap_scores)


def evaluate(qrels, results):
    """Evaluate retrieval results"""
    print(f"\n📈 Evaluating...")
    
    metrics = {
        "ndcg@1": calculate_ndcg(qrels, results, 1),
        "ndcg@5": calculate_ndcg(qrels, results, 5),
        "ndcg@10": calculate_ndcg(qrels, results, 10),
        "recall@1": calculate_recall(qrels, results, 1),
        "recall@5": calculate_recall(qrels, results, 5),
        "recall@10": calculate_recall(qrels, results, 10),
        "precision@1": calculate_precision(qrels, results, 1),
        "precision@5": calculate_precision(qrels, results, 5),
        "precision@10": calculate_precision(qrels, results, 10),
        "map@10": calculate_map(qrels, results, 10),
    }
    
    return metrics


def benchmark_index(index_type, corpus_data, query_data, qrels):
    """Benchmark a single OpenSearch index type"""
    print(f"\n{'='*60}")
    print(f"Testing: {index_type}")
    print(f"{'='*60}")
    
    index_name = f"finder_{index_type.lower()}"
    
    try:
        # Connect to OpenSearch
        client = get_opensearch_client()
        
        # Verify connection
        info = client.info()
        print(f"✅ Connected to OpenSearch {info['version']['number']}")
        
        # Create index
        create_index(client, index_name, index_type)
        
        # Insert corpus
        insert_corpus(client, index_name, corpus_data)
        
        # Build index and wait
        build_time = build_index_and_wait(client, index_name, index_type)
        
        # Search queries
        results, search_time, avg_query_time = search_queries(
            client, index_name, query_data, index_type
        )
        
        # Evaluate
        metrics = evaluate(qrels, results)
        
        # Prepare output
        output = {
            "index_type": index_type,
            "database": "OpenSearch",
            "index_build_time_sec": round(build_time, 2),
            "total_search_time_sec": round(search_time, 2),
            "avg_query_time_ms": round(avg_query_time * 1000, 2),
            **{k: round(v, 4) for k, v in metrics.items()}
        }
        
        # Save individual result
        os.makedirs("./results", exist_ok=True)
        output_file = f"./results/opensearch_results_{index_type.lower()}.json"
        with open(output_file, 'w') as f:
            json.dump(output, f, indent=2)
        
        print(f"\n✅ {index_type} Complete!")
        print(f"   NDCG@10: {metrics['ndcg@10']:.4f}")
        print(f"   Recall@10: {metrics['recall@10']:.4f}")
        print(f"   Precision@10: {metrics['precision@10']:.4f}")
        print(f"   MAP@10: {metrics['map@10']:.4f}")
        
        return output
        
    except Exception as e:
        print(f"❌ ERROR with {index_type}: {e}")
        import traceback
        traceback.print_exc()
        return None


def display_results(opensearch_results):
    """Display OpenSearch results"""
    print("\n" + "="*100)
    print("OPENSEARCH k-NN BENCHMARK SUMMARY")
    print("="*100)
    
    df = pd.DataFrame(opensearch_results)
    
    # Key metrics
    print("\n📊 KEY PERFORMANCE METRICS:")
    print("-"*100)
    summary_df = df[['index_type', 'index_build_time_sec', 'total_search_time_sec',
                      'avg_query_time_ms', 'ndcg@10', 'recall@10', 'precision@10', 'map@10']]
    print(summary_df.to_string(index=False))
    
    # Best performers
    print("\n\n🏆 BEST PERFORMERS:")
    print("-"*100)
    print(f"Fastest Search:      {df.loc[df['avg_query_time_ms'].idxmin(), 'index_type']} "
          f"({df['avg_query_time_ms'].min():.2f}ms per query)")
    print(f"Highest NDCG@10:     {df.loc[df['ndcg@10'].idxmax(), 'index_type']} "
          f"({df['ndcg@10'].max():.4f})")
    print(f"Highest Recall@10:   {df.loc[df['recall@10'].idxmax(), 'index_type']} "
          f"({df['recall@10'].max():.4f})")
    
    return df


def main():
    """Run all OpenSearch benchmarks"""
    print("="*60)
    print("OPENSEARCH k-NN BENCHMARKING SUITE")
    print("="*60)
    print("\n📌 Reusing embeddings from Milvus encoding step!")
    print(f"📋 Will test {len(INDEX_CONFIGS)} index types:")
    for idx, config in INDEX_CONFIGS.items():
        print(f"   - {idx}: {config['description']}")
    
    # Check OpenSearch connection
    try:
        client = get_opensearch_client()
        info = client.info()
        print(f"\n✅ OpenSearch connection OK")
        print(f"   Version: {info['version']['number']}")
        print(f"   Cluster: {info['cluster_name']}")
    except Exception as e:
        print(f"\n❌ ERROR: Cannot connect to OpenSearch: {e}")
        print("   Make sure OpenSearch is running with k-NN plugin enabled")
        return
    
    # Load data
    corpus_data, query_data = load_embeddings()
    qrels = load_qrels()
    print(f"✅ Loaded {len(qrels)} query judgments")
    
    # Run benchmarks
    all_results = []
    
    for index_type in INDEX_CONFIGS.keys():
        result = benchmark_index(index_type, corpus_data, query_data, qrels)
        if result:
            all_results.append(result)
        
        # Wait between tests
        if index_type != list(INDEX_CONFIGS.keys())[-1]:
            print("\n⏳ Waiting 5 seconds before next test...")
            time.sleep(5)
    
    # Save combined results
    with open("./results/opensearch_combined_results.json", 'w') as f:
        json.dump(all_results, f, indent=2)
    
    print("\n💾 All OpenSearch results saved to ./results/")
    
    # Display comparison
    df = display_results(all_results)
    
    print("\n\n🎉 All OpenSearch benchmarks complete!")
    
    return df


if __name__ == "__main__":
    results_df = main()

OPENSEARCH k-NN BENCHMARKING SUITE

📌 Reusing embeddings from Milvus encoding step!
📋 Will test 5 index types:
   - NO_INDEX: Exact search using script scoring (like Milvus FLAT)
   - FAISS_HNSW: FAISS engine with HNSW algorithm
   - FAISS_IVF: FAISS engine with IVF algorithm
   - NMSLIB_HNSW: nmslib engine with HNSW algorithm
   - LUCENE_HNSW: Lucene engine with HNSW algorithm

✅ OpenSearch connection OK
   Version: 2.11.0
   Cluster: docker-cluster

📥 Loading pre-encoded embeddings...
✅ Loaded:
   Corpus: (13867, 1024)
   Queries: (216, 1024)
✅ Loaded 64 query judgments

Testing: NO_INDEX
✅ Connected to OpenSearch 2.11.0
❌ ERROR with NO_INDEX: RequestError(400, 'mapper_parsing_exception', 'No handler for type [dense_vector] declared on field [embedding]')

⏳ Waiting 5 seconds before next test...


/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_54315/1989178554.py:158: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  qrels = df.groupby("query_id").apply(
Traceback (most recent call last):
  File "/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_54315/1989178554.py", line 505, in benchmark_index
    create_index(client, index_name, index_type)
  File "/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_54315/1989178554.py", line 210, in create_index
    client.indices.create(index=index_name, body=index_body)
  File "/opt/homebrew/anaconda3/envs/FinRAG/lib/python3.11/site-packages/opensearchpy/client/utils.py", line 176, in _wrapped
    if p in kwargs:
       ^^^^^^^^^^^


Testing: FAISS_HNSW
✅ Connected to OpenSearch 2.11.0
❌ ERROR with FAISS_HNSW: RequestError(400, 'mapper_parsing_exception', 'Failed to parse mapping [_doc]: Validation Failed: 1: "hnsw" configuration does not support space type: "cosinesimil".;')

⏳ Waiting 5 seconds before next test...


Traceback (most recent call last):
  File "/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_54315/1989178554.py", line 505, in benchmark_index
    create_index(client, index_name, index_type)
  File "/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_54315/1989178554.py", line 210, in create_index
    client.indices.create(index=index_name, body=index_body)
  File "/opt/homebrew/anaconda3/envs/FinRAG/lib/python3.11/site-packages/opensearchpy/client/utils.py", line 176, in _wrapped
    if p in kwargs:
       ^^^^^^^^^^^^^
  File "/opt/homebrew/anaconda3/envs/FinRAG/lib/python3.11/site-packages/opensearchpy/client/indices.py", line 244, in create
    :arg include_defaults: Whether to return all default setting for
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/anaconda3/envs/FinRAG/lib/python3.11/site-packages/opensearchpy/transport.py", line 457, in perform_request
    if data:
        ^^^^^
  File "/opt/homebrew/anaconda3/en


Testing: FAISS_IVF
✅ Connected to OpenSearch 2.11.0
❌ ERROR with FAISS_IVF: RequestError(400, 'mapper_parsing_exception', 'Failed to parse mapping [_doc]: Validation Failed: 1: "method" requires training.;2: "ivf" configuration does not support space type: "cosinesimil".;')

⏳ Waiting 5 seconds before next test...


Traceback (most recent call last):
  File "/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_54315/1989178554.py", line 505, in benchmark_index
    create_index(client, index_name, index_type)
  File "/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_54315/1989178554.py", line 210, in create_index
    client.indices.create(index=index_name, body=index_body)
  File "/opt/homebrew/anaconda3/envs/FinRAG/lib/python3.11/site-packages/opensearchpy/client/utils.py", line 176, in _wrapped
    if p in kwargs:
       ^^^^^^^^^^^^^
  File "/opt/homebrew/anaconda3/envs/FinRAG/lib/python3.11/site-packages/opensearchpy/client/indices.py", line 244, in create
    :arg include_defaults: Whether to return all default setting for
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/anaconda3/envs/FinRAG/lib/python3.11/site-packages/opensearchpy/transport.py", line 457, in perform_request
    if data:
        ^^^^^
  File "/opt/homebrew/anaconda3/en


Testing: NMSLIB_HNSW
✅ Connected to OpenSearch 2.11.0
✅ Created index: finder_nmslib_hnsw

📤 Inserting 13867 documents into OpenSearch...
   Prepared 5000/13867 documents
   Prepared 10000/13867 documents
✅ Inserted 13867 documents (failed: 0)
✅ Index refreshed

🔨 Building NMSLIB_HNSW index...
   Description: nmslib engine with HNSW algorithm
✅ Index built in 5.71s

🔍 Searching 216 queries...
   Method: k-NN Index
   Processed 50/216 queries (avg: 31.40ms/query)
   Processed 100/216 queries (avg: 26.41ms/query)
   Processed 150/216 queries (avg: 27.85ms/query)
   Processed 200/216 queries (avg: 28.08ms/query)

✅ Search complete:
   Total time: 6.09s
   Avg per query: 28.19ms

📈 Evaluating...

✅ NMSLIB_HNSW Complete!
   NDCG@10: 0.4183
   Recall@10: 0.5586
   Precision@10: 0.0781
   MAP@10: 0.3607

⏳ Waiting 5 seconds before next test...

Testing: LUCENE_HNSW
✅ Connected to OpenSearch 2.11.0
✅ Created index: finder_lucene_hnsw

📤 Inserting 13867 documents into OpenSearch...
   Prepared